src/lib/mina_net2/libp2p_helper.ml

```diff
@@ -0,0 +1,288 @@
+open Async
+open Core
+open Pipe_lib
+open O1trace
+
+exception Libp2p_helper_died_unexpectedly
+
+module Go_log = struct
+  let ours_of_go lvl =
+    let open Logger.Level in
+    match lvl with
+    | "error" | "panic" | "fatal" ->
+        Error
+    | "warn" ->
+        Debug
+    | "info" ->
+        (* this is intentionally debug, because the go info logs are too verbose for our info *)
+        Debug
+    | "debug" ->
+        Spam
+    | _ ->
+        Spam
+
+  (* there should be no other levels. *)
+
+  type record =
+    { ts : string
+    ; module_ : string [@key "logger"]
+    ; level : string
+    ; msg : string
+    ; metadata : Yojson.Safe.t String.Map.t
+    }
+
+  let record_of_yojson (json : Yojson.Safe.t) =
+    let open Result.Let_syntax in
+    let prefix = "Mina_net2.Go_log.record_of_yojson: " in
+    match json with
+    | `Assoc fields ->
+        let set_field field_name prev_value parse json =
+          match prev_value with
+          | Some _ ->
+              Error
+                (prefix ^ "Field '" ^ field_name ^ "' appears multiple times")
+          | None ->
+              parse json
+              |> Result.map_error ~f:(fun err ->
+                     prefix ^ "Could not parse field '" ^ field_name ^ "':"
+                     ^ err)
+              |> Result.map ~f:Option.return
+        in
+        let get_field field_name value =
+          match value with
+          | Some x ->
+              Ok x
+          | None ->
+              Error (prefix ^ "Field '" ^ field_name ^ "' is required")
+        in
+        let string_of_yojson = function
+          | `String s ->
+              Ok s
+          | _ ->
+              Error "Expected a string"
+        in
+        let%bind ts, module_, level, msg, metadata =
+          List.fold_result ~init:(None, None, None, None, String.Map.empty)
```

Comment:
It would be cleaner to put all fields to the map and then access all the fields through `map`, removing them after access